services/device-registry-file/src/main/java/org/eclipse/hono/deviceregistry/file/FileBasedCredentialsService.java

```diff
@@ -608,15 +609,25 @@ private boolean hasUniqueSecretIds(final JsonObject credentialObject) {
                 .map(CommonCredential::stripPrivateInfo)
                 .collect(Collectors.toList());
 
+        // resolve cache directive for results.
+        // Limit caching to the case when cache directives are the same for all results
+        final Set<CacheDirective> cacheDirectives = result.stream().map(CommonCredential::getType)
```

Comment:
FMPOV we do not need to support caching of this response as it is part of the Registry Management API which we expect to be invoked rarely only. The CredentialsService, on the other hand, is being invoked permanently by the protocol adapters during runtime. Thus it is more important to support caching in the adapters.

Reply:
In the tests I was doing, this was causing a bottleneck. I have improved the change in the most recent commits. Please re-review.

Reply:
Would you mind elaborating a little on what you had been doing in your tests and why you believe that the fact that the registry doesn't include a cache directive in responses to the management API operations caused a bottleneck?

Reply:
@sophokles73 The reason why I started making changes to cache `CredentialsManagementService.readCredentials` calls was because of the TODO that was present in FileBasedCredentialService.processReadCredentials method: https://github.com/eclipse/hono/blob/251b1472f0275b2174ea07d9ec00bd66d55fd4d0/services/device-registry-file/src/main/java/org/eclipse/hono/deviceregistry/file/FileBasedCredentialsService.java#L611-L620 .

Adding caching headers to CredentialsManagementService.readCredentials calls most likely didn't fix the performance bottleneck I was experiencing. IIRC, it was the fix (also part of this PR) to add caching headers to `TenantManagementService.readTenant`. I assume that there's a bug somewhere that the TenantManagementService.readTenant gets called on the "hot path", when new telemetry is being received. I haven't had the chance to verify my assumptions.

Reply:
All implementations of the Tenant service include a proper max-age cache directive in its responses. So that should not be your problem. However, the JDBC based Credentials service implementation always includes a no-cache directive which prevents the protocol adapters from caching credentials. Depending on the kind of (device) protocol you have been using in your tests, this could have a major impact. For instance, when using the HTTP protocol adapter, each request from a device needs to verify the device's credentials again. So if your device used Basic Auth for authentication, the credentials are being retrieved from the Credentials service for each request.
Based on that I would very much like to accept your changes adding the cache directive to the JDBC based Credentials service implementation. But I would still rather not add cache directives to the responses returned by the Registry Management API. I can understand if you do not consider making this change worth your time anymore. In that case I will be happy to add the missing code to the JDBC registry myself. I just wanted to give you the chance to get your code merged ... 
Let me know how you want to proceed. Thanks in any case for your interest in Hono and your contributions so far :+1:

Reply:
@lhotari any opinion? How should we proceed?

Reply:
I have started a new job and I don't have time to work on this PR. I'll close this PR.

Reply:
I see. Thanks for your contributions and good luck and all the best for your new job :+1:

Reply:
Thanks @sophokles73